csrc/preseg_passes/translate_repeat_to_expand.cpp

```diff
@@ -0,0 +1,194 @@
+      }
+
+      auto cat_op = pad_out->uses().at(0);
+
+      // If other inputs to the same concat op are already found, make
+      // sure this path from the pad op is compatible with the known
+      // ops.
+      if (auto it = repeat_info_map_.find(cat_op);
+          it == repeat_info_map_.end()) {
+        RepetitionInfo info;
+        info.input_tv = pad_inp;
+        info.repeated_id = inp_padded_id;
+        info.cat_inp_tvs.push_back(pad_out);
+        repeat_info_map_.emplace(cat_op, info);
+      } else {
+        auto& info = repeat_info_map_.at(cat_op);
+        if (info.input_tv != pad_inp || info.repeated_id != inp_padded_id) {
+          // Invalid
+          repeat_info_map_.erase(cat_op);
+          continue;
+        }
+        info.cat_inp_tvs.push_back(pad_out);
+      }
+    }
+
+    // Remove invalid entries
+    for (auto it = repeat_info_map_.begin(); it != repeat_info_map_.end();) {
+      Expr* concatenating_expr = it->first;
+      const RepetitionInfo& info = it->second;
+      // Make sure all inputs to concatenating_expr are detected
+      if (concatenating_expr->inputs().size() != info.cat_inp_tvs.size()) {
+        // Invalid
+        it = repeat_info_map_.erase(it);
+        continue;
+      }
+      ++it;
+    }
+  }
+
+  // For each detected repetition:
+  //
+  // Step 1. Insert a broadcast ID immediately outside of the
+  // repeated ID
+  // Step 2. Expand the broadcast ID by the repetition factor
+  // Step 3. Flatten the expanded ID and the repeated ID
+  void translate() {
+    const auto exprs = fusion_->exprs();
+    // Apply the translation in a reverse topological order. Since the
+    // output of the repetition is replaced, the use exprs of the
+    // output are replaced too, which may invalidate the inspected
+    // info invalid.
+    for (auto exprs_it = exprs.rbegin(); exprs_it != exprs.rend(); ++exprs_it) {
+      Expr* expr = *exprs_it;
+      auto repeat_info_map_it = repeat_info_map_.find(expr);
+      if (repeat_info_map_it == repeat_info_map_.end()) {
+        continue;
+      }
+
+      const auto& info = repeat_info_map_it->second;
+
+      if (info.cat_inp_tvs.size() < 2) {
+        continue;
+      }
+
+      auto original_out_tv = expr->output(0)->as<TensorView>();
+
+      // Step 1
+      auto inp_domain =
+          TensorDomain::noReductions(info.input_tv->getLogicalDomain());
+      std::vector<bool> bcast_flags(inp_domain.size() + 1, false);
+      auto repeated_id_offset = std::distance(
+          inp_domain.begin(),
+          std::find(inp_domain.begin(), inp_domain.end(), info.repeated_id));
+      bcast_flags.at(repeated_id_offset) = true;
+      auto broadcast_tv = broadcast(info.input_tv, bcast_flags);
+      NVF_ERROR(broadcast_tv->nDims() == inp_domain.size() + 1);
+
+      // Step 2
+      std::vector<Val*> expanded_sizes(
+          bcast_flags.size(), IrBuilder::create<Val>(-1L));
+      expanded_sizes.at(repeated_id_offset) = IrBuilder::create<Val>(2L);
```

Comment:
This shouldn't be a hard-coded 2 here. I think our logic can handle something like `cat({t0, t0, t0}, -1)` as well and we don't have a check to exclude that neither.
We should use the # of inputs here instead.

Reply:
^^^ Let's also add a test case with that.

Reply:
Ah, that's a very good point. Yes, I took a shortcut and forgot to update it.